Add stable release policy (#685)
cosmos/ibc-go: RELEASES.md

We ensure all major releases are supported by relayers ([hermes](https://github.com/informalsystems/ibc-rs), [rly](https://github.com/strangelove-ventures/relayer) and [ts-relayer](https://github.com/confio/ts-relayer) at the moment) which can relay between the new major release and older releases. We have no plans of upgrading to an IBC protocol specification v2.0, as this would be very disruptive to the ecosystem.

## Stable Release Policy

> A Major Release series continues to receive bug fixes (released as a Patch Release) until it reaches End Of Life. A Major Release series reaches End of Life 6 months after the publication of the next Major Release series. For example, if the current Major Release series is v1, then v1 will be supported for up to 6 months after the publication of v2. 

Reviewer: I personally would just use lowercase words, but I am not super strongly against otherwise...
```suggestion
A major release series continues to receive bug fixes (released as a patch release) until it reaches End Of Life. A major release series reaches End of Life 6 months after the publication of the next major release series. For example, if the current major release series is v1, then v1 will be supported for up to 6 months after the publication of v2. 
```

> A major release series continues to receive bug fixes (released as a patch release) until it reaches End Of Life. A major release series reaches End of Life 1 year after its initial release. An extension to the End of Life date is made if the next major release is not made for 6 months. In this situation, the End of Life is the date of publication for the next major release plus 6 months. 

Reviewer 2: Since using lower casing for "major release" as @crodriguezvega suggested, should we also update "End of Life"?

```suggestion
A major release series continues to receive bug fixes (released as a patch release) until it reaches end of life. A major release series reaches end of life 1 year after its initial release. An extension to the end of life date is made if the next major release is not made for 6 months. In this situation, the end of life is the date of publication for the next major release plus 6 months. 
```

Author: maybe we should use [dashes](https://en.wikipedia.org/wiki/End-of-life_product)?

Reviewer 2: I actually agree, it seems to be [common format](https://en.wikipedia.org/wiki/End-of-life_product)

> A major release series continues to receive bug fixes (released as a patch release) until it reaches end of life. A major release series reaches end of life 1 year after its initial release. An extension to the end of life date is made if the next major release is not made for 6 months. In this situation, the end of life is the date of publication for the next major release plus 6 months. 

Reviewer: So when we talk here about major release we refer to all the release series in, for example, v1 (i.e. 1.0.x, 1.1.x, etc), right?

Author: yes

Author: Is there a sentence you'd like me to add clarifying this?

> A major release series continues to receive bug fixes (released as a patch release) until it reaches end of life. A major release series reaches end of life 1 year after its initial release. An extension to the end of life date is made if the next major release is not made for 6 months. In this situation, the end of life is the date of publication for the next major release plus 6 months. 

Reviewer: Could there be some confusion here with the term major release? We say that v1 is a major release and that is going to reach end of life after a year, but for example:
- v1.0.x is created on January 1st 2021.
- v1.1.x is created on July 1st 2021.
- v2.0.x is created on February 1st 2022.
Then v1.0.x reaches end of life on January 1st, 2022. But then v1.1.x will reach end of life August 1st , 2022? However this is still a v1 major release, right?

Author: I think we are getting lost in the weeds. Lets think about the user perspective and why they want support for minor vs major releases. 

Updating to a different minor release implies a coordinated chain upgrade. 
Updating to a different major release implies a code upgrade + coordinated chain upgrade. 

We want to give developers ample time to update their code to the next version so we don't disrupt their workflow. 

The 6 month policy is basically saying, you have 6 months to update your code once the next major release is done. If we release the next major release very fast, you have a little more than 6 months. 

So once v2 is released, the 6 month timer should be applied to all existing release lines

How about this. We define a release line as either major or minor. When the next major release hasn't been made, the end of life of the release line is 1 year (with the possibility of extension). When the next major release has already been made, the end of life is 6 months. 

So in the above example, the end of life for v1.2.x is October 1st, 2022

Author: It is possible this policy is going to force us to maintain release lines for a long time, but there's only one way to find out :slightly_smiling_face: 

Thoughts?

Reviewer: What do you think of these changes?

```suggestion
The beginning of a new major release is marked by the release of a new major version. A major release series is comprised of all minor and patch releases made under the same major version number. The series continues to receive bug fixes (released as minor or patch releases) until it reaches end of life. The date when a major release series reaches end of life is determined by one of the two following methods:
- If the next major release is made within the first 6 months, then the end of life date of the major release series is 1 year after its initial release. 
- If the next major release is made 6 months after the initial release, then the end of life date of the major release series is 6 months after the release date of the next major release.
```

Author: Looks good! I guess the table needs to be updated then? Should v1 reach end of life a year from merging this pr?

Reviewer: Should we say maybe that v1 gets 6 months and v2 one year?

Author: Yea that sounds good to me! I don't really expect many folks to be on v1 since many are going from stargate -> v2 since v1 wasn't out by itself for too long

> For example, if the current major release series is v1 and was released on January 1st, 2022, then v1 will be supported at least until January 1st, 2023. If v2 is published on August 1st 2022, then v1's End of Life will be March 1st, 2023. 

Reviewer: So if we didn't make a new major release (for example, v2) after the previous one (for example, v1) for longer than a year, does that mean that the previous major release (v1) reaches end of life the moment the new major release is released?

> For example, if the current major release series is v1 and was released on January 1st, 2022, then v1 will be supported at least until January 1st, 2023. If v2 is published on August 1st 2022, then v1's End of Life will be March 1st, 2023. 

Author: No, v1 would reach end of life 6 months after v2 is released

|Release|End of Life Date|
|-------|-------|
|`v1.1.x`|July 05, 2022|
|`v1.2.x`|July 05, 2022|
|`v2.0.x`|6 months after v3 release|

> At the time this policy was adopted, v1 and v2 had already been released. The 6 month policy for the v1 Major Release series will go into affect on the day this policy was merged into the codebase as opposed to when v2 was released.

Reviewer: ```suggestion
At the time this policy was adopted, v1 and v2 had already been released. The 6 month policy for the v1 Major Release series will go into effect on the day this policy was merged into the codebase as opposed to when v2 was released.
```
